include/picongpu/plugins/xrayScattering/XrayScattering.hpp

```diff
@@ -0,0 +1,846 @@
+            auto tmpField = dc.get< FieldTmp >(
+                FieldTmp::getUniqueId( 0 ),
+                true
+            );
+            // Initiate with zeros.
+            tmpField->getGridBuffer( ).getDeviceBuffer( ).setValue(
+                    FieldTmp::ValueType::create( 0.0 ) );
+            // Chose species.
+            auto species = dc.get< T_ParticlesType >(
+                T_ParticlesType::FrameType::getName( ),
+                true
+            );
+
+            // Chose the solver for populating a TmpField with the form factor
+            // density of the particles.
+            using ElectronDensitySolver =
+                typename DetermineElectronDensitySolver<
+                T_ParticlesType >::type;
+            // Calculate density.
+            tmpField-> template computeValue<
+                CORE + BORDER,
+                ElectronDensitySolver
+            >(
+                * species,
+                currentStep
+            );
+            // Release particle data.
+            dc.releaseData( T_ParticlesType::FrameType::getName( ) );
+            // Get the field data box.
+            FieldTmp::DataBoxType tmpFieldBox =
+                tmpField->getGridBuffer( ).getDeviceBuffer( ).getDataBox( );
+            return tmpFieldBox;
+        }
+
+
+        /** Runs xrayScattering kernel when the output is split over nodes.
+         *
+         * A single kernel run adds result only to that output part which
+         * currently resides on the node. The output parts are passed along to
+         * the neigbouring node, in a circle. This repeats until every node has
+         * computed all scattering vectors.
+         *
+         * @param cellsGrid field grid, without GUARD, on one device
+         * @param fieldTmpNoGuard field data
+         * @param globalOffset offset from the global to the local domain
+         * @param numBlocks number of virtual blocks used in a kernel run
+         * @param fieldPos TmpField in cell position
+         */
+        template < typename T_FieldPos >
+        HINLINE void runKernelInSplitMode(
+            DataSpace < simDim > & cellsGrid,
+            FieldTmp::DataBoxType const & fieldTmpNoGuard,
+            DataSpace< simDim > & globalOffset,
+            uint32_t const & numBlocks,
+            T_FieldPos const & fieldPos
+        )
+        {
+            // The available number of virtual workers.
+            constexpr uint32_t numWorkers = pmacc::traits::GetNumWorkers<
+                pmacc::math::CT::volume< SuperCellSize >::type::value >::value;
+
+            uint32_t countVectors, iterOffset;
+            // Loop over kernel runs.
+            for( uint32_t step = 0; step < countRanks; step++ )
+            {
+                // Pass along the data.
+                communicationOnStep( step );
+                // 1D offset to the begin of the currently processed output
+                // part.
+                iterOffset = calcOffset( step );
```

Comment:
Maybe `iterOffset` can be defined here, as it seems independent for all loop iterations,

Reply:
ahh yes I think, thought at some point that I' cant have declarations in loops, for whatever reason :smile:.